including px-vis-markers in px-vis-xy-chart

saiDevarakonda opened the pull request:
# Pull Request

Hi,

Thanks for helping us improve the Predix UI platform by submitting a Pull Request.

To help us merge your Pull Request as fast as possible, please fill out the following sections:

* ## A description of the changes proposed in the pull request: Including px-vis-markers in px-vi-xy-chart

* ## A reference to a related issue (if applicable): Allow makers in in px-vis-xy-chart similar to px-vis-timeseries.

* ## @mentions of the person or team responsible for reviewing proposed changes:

* ## working tests:
#### The tests need to be functional and/or unit tests, written for the wct framework, and placed in the test folder, following our testing guidelines.

saiDevarakonda committed:
including px-vis-markers in px-vis-xy-chart

evanjd commented on px-vis-xy-chart.html:
```diff
@@ -607,10 +623,33 @@
+       */
+      marginMarkersAdjustment: {
+        type: Object,
+        value: {},
```
This needs to be a function that returns an empty object, otherwise this property will be shared across all instances of this component.

See here for more detail: https://polymer-library.polymer-project.org/1.0/docs/devguide/properties#configure-values

```suggestion
        value: function() { return {}; }
```

saiDevarakonda replied:
Done: Updated changes in new commit.

evanjd commented on px-vis-xy-chart.html:
```diff
@@ -851,6 +892,39 @@
+      if(this.hasUndefinedArguments(arguments)) {
+        return;
+      }
+
```
Can you please cleanup the whitespace here? There should be no more than one consecutive newline between code blocks.

saiDevarakonda replied:
Done: Updated changes in new commit.

evanjd commented on px-vis-xy-chart.html:
```diff
@@ -851,6 +892,39 @@
+
+
+      //use our internal margins
+      newMarg = {
```
```suggestion
      var newMarg = {
```

saiDevarakonda replied:
Done: Updated changes in new commit.

evanjd commented on px-vis-xy-chart.html:
```diff
@@ -851,6 +892,39 @@
+      }
+
+
+      var newMarg;
```
No need to separate the declaration from the initialization, remove this and add `var` to line 905.

saiDevarakonda replied:
Done: Updated changes in new commit.

evanjd commented on px-vis-xy-chart.html:
```diff
@@ -851,6 +892,39 @@
+      }
+
+      //adjust if something is dev set
+      newMarg.top += this.marginMarkersAdjustment.top  ? Number(this.marginMarkersAdjustment.top) : 0;
```
Tidy up whitespace here please. This can also probably be shortened to:
```suggestion
      newMarg.top += Number(this.marginMarkersAdjustment.top || 0);
```

saiDevarakonda replied:
Done: Updated changes in new commit.

evanjd reviewed (changes requested):
Thanks for the contribution. _Please_ provide an example of what your change does, preferably with sample code.

evanjd commented:
@cla-bot check

saiDevarakonda committed:
updating PR Comments

evanjd reviewed (changes requested):
Thanks for the updates, but I still need more information to understand your change. A screenshot and some sample code that demonstrates what you've added would be a good start.

priyabrat1801 commented:
@evanjd the PR in px-vis (https://github.com/predixdesignsystem/px-vis/pull/151) is related as well. This is similar to a feature existing in px-vis-timeseries for event markers which is needed in xy-chart component for APM Analysis.

Is it ok to do a demo on skype with @saiDevarakonda to go over the feature.

saiDevarakonda commented:
HI Evan,
Sent you and email with screenshot and sample data @ evan.bruhn@ge.com

evanjd commented:
> Is it ok to do a demo on skype with @saiDevarakonda to go over the feature.

Appreciate the offer but I'd be happy if you simply provide a minimal explanation and example of your code that demonstrates your change, something like:

"Hi, I added [[feature]] to solve [[problem]].

This is what it looked like before.
[[screenshot]]

This is what it looks like now.
[[screenshot]].

Here is some sample code that demonstrates it

```html
<px-vis-xy-chart
   whatever-props-are-needed-go-here
/>
```
"

saiDevarakonda committed:
new property xAxisLocation for px-vis-markers

saiDevarakonda commented:
"Hi, I added [[px-vis-markers]] to show [[px-vis-markers]] on px-vis-xy-chart.

This is what it looked like before.
```<px-vis-xy-chart id="pxChart_{{chart.id}}" render-to-canvas="true"
												   show-gaps
												   height="[[smartSignalchartHeight]]"
												   width="{{ xyChartWidth }}"
												   margin="[[ chartMargin ]]"
												   disable-navigator
												   horizontal-line="none"
												   default-series-config='{
												  "type": "line",
												  "markerSymbol": "diamond",
												  "markerSize": "16",
												  "markerFillOpacity": "0.1",
												  "x": "TimeStamp"
												}'
												   toolbar-config="[[ _getSsToolbarConfig(isCrosshairFrozen) ]]"
												   time-data="x"
												   time-domain="[[ selectedDomain ]]"
												   y-axis-config="[[ yAxisConfig ]]"
												   register-config="[[ ssRegisterConfig ]]"
												   chart-data="[[xyVisChartData]]"
												   chart-extents="[[ _chartExtents ]]"
												   threshold-config="[[ thresholdConfig ]]"
												   display-threshold-title
												   threshold-data="[[ thresholdData ]]"
												   series-config="{{seriesConfigObjs}}"
												   muted-series="{{ mutedSeries }}"
												   tooltip-config="[[ssTooltipConfig]]"
												  

												   highlighter-config="[[ highlighterConfig ]]"
												   >
								</px-vis-xy-chart>
```
[[screenshot]]
![image](https://user-images.githubusercontent.com/7873474/54159036-3a98d600-4409-11e9-87cd-81f86ea24d95.png)

This is what it looks like now.
ADDED markerData,markerConfig.

```<px-vis-xy-chart id="pxChart_{{chart.id}}" render-to-canvas="true"
												   show-gaps
												   height="[[smartSignalchartHeight]]"
												   width="{{ xyChartWidth }}"
												   margin="[[ chartMargin ]]"
												   disable-navigator
												   horizontal-line="none"
												   default-series-config='{
												  "type": "line",
												  "markerSymbol": "diamond",
												  "markerSize": "16",
												  "markerFillOpacity": "0.1",
												  "x": "TimeStamp"
												}'
												   toolbar-config="[[ _getSsToolbarConfig(isCrosshairFrozen) ]]"
												   time-data="x"
												   time-domain="[[ selectedDomain ]]"
												   y-axis-config="[[ yAxisConfig ]]"
												   register-config="[[ ssRegisterConfig ]]"
												   chart-data="[[xyVisChartData]]"
												   chart-extents="[[ _chartExtents ]]"
												   threshold-config="[[ thresholdConfig ]]"
												   display-threshold-title
												   threshold-data="[[ thresholdData ]]"
												   series-config="{{seriesConfigObjs}}"
												   muted-series="{{ mutedSeries }}"
												   tooltip-config="[[ssTooltipConfig]]"
												   marker-config="[[ markerConfig ]]"
												   marker-data="[[ eventData ]]"
												   first-date-time-format="[[firstDateTimeFormat]]"
												   timezone="[[timezone]]"
												   highlighter-config="[[ highlighterConfig ]]"
												   >
								</px-vis-xy-chart>
```
[[screenshot]].

![image](https://user-images.githubusercontent.com/7873474/54159077-57cda480-4409-11e9-9385-5904e0955280.png)

Here is some sample code that demonstrates it

Sample MarkerData
```

[

{

"id": 1,

"label": "Advisory",

"time": 1505433900000,

"datapoint": 1

},

{

"id": 2,

"label": "Advisory",

"time": 1508025900000,

"datapoint": 2

},

{

"id": 3,

"label": "Advisory",

"time": 1510704300000,

"datapoint": 3

},

{

"id": 4,

"label": "Advisory",

"time": 1510963500000,

"datapoint": 4

},

{

"id": 5,

"label": "Advisory",

"time": 1511546100000,

"datapoint": 5

},

{

"id": 1,

"label": "Adaptation-Successful",

"time": 1505433900000,

"datapoint": 1

},

{

"id": 2,

"label": "Outlier",

"time": 1505434500000,

"datapoint": 2

},

{

"id": 2,

"label": "VSG",

"time": 1505434500000,

"datapoint": 2

},

{

"id": 3,

"label": "Adaptation-Successful",

"time": 1508026200000,

"datapoint": 3

}

]
```
Sample Marker Config
```
{

"Advisory": {

"location": "top",

"color": "rgb(180,140,242)",

"markerSymbol": "circle",

"priority": 4,

"markerSize": 25,

"markerScale": 1,

"markerFillOpacity": "1",

"markerStrokeOpacity": 1,

"row": 0,

"timezone": "UTC",

"firstDateTimeFormat": "hh:mm:ss A z"

},

"Adaptation-Successful": {

"location": "bottom",

"color": "rgb(222,82,136)",

"markerSymbol": "circle",

"priority": 5,

"markerSize": 34,

"markerScale": 1,

"markerFillOpacity": "1",

"markerStrokeOpacity": 1,

"row": 2,

"timezone": "UTC",

"firstDateTimeFormat": "hh:mm:ss A z"

}

}
```

gedigital-apm commented:
@evanjd could you please check the latest and see if its ok to merge. We are closing out the release in Q1 and need this to be included. Appreciate the help.

evanjd commented:
@gedigital-apm @saiDevarakonda I'm looking for a **_minimal_** example implementation that demonstrates your change. Your example will not work in isolation as it had bindings to other variables not in your sample code.

Something really simple is all I'm looking for. Don't include any configuration that doesn't need to be there.

Also can you [sign the CLA](https://www.predix-ui.com/#/develop/cla)? I've fixed the bug in the form, it should work for you now.

saiDevarakonda commented:
Hi Evan updated the sample code above. Please review. Thank you.

evanjd commented:
@cla-bot check

evanjd reviewed (changes requested):
Your example implementation doesn't work. This is what it looks like (and yes I've pulled in your copy of px-vis into my instance of px-vis-xy-chart component).

![Screen Shot 2019-03-20 at 11 49 58 am](https://user-images.githubusercontent.com/4055224/54651307-6e10ea00-4b06-11e9-865d-21466db64ca1.png)


It would help if you kept your example as simple as possible like I said earlier. A small, contrived example implementation that clearly demonstrates your change is what we want.

Also, you haven't explained what this is for. When I provided this example:

> Hi, I added [[feature]] to solve [[problem]].

I was prompting you for a simple explanation of what you added and why. This isn't helpful:

> Hi, I added [[px-vis-markers]] to show [[px-vis-markers]] on px-vis-xy-chart.

I'm looking for something like: "At APM, we have a need to plot data across a secondary axis outside the chart viewport. In our case, we're plotting [[something]] for [[whatever reason]]". I'm just making this up but hopefully you get the idea. Just a simple of explanation of what you're doing and why.

Also, it would probably be a good idea to update the demo to include the properties you've added.

saiDevarakonda committed:
updating the demo

saiDevarakonda commented:
**APM Requirement :** 

Based in customer need, we need to plot SS charts by points and the chart X-axis should scale based on points instead of timestamp.

**Acceptance Criteria .**

 **X-Y chart** can be used to replace the **ts chart** when user is in points mode.
 All current **markers** configuration in **ts chart** should be supported in the **X-Y chart**.
________________________________________________________________________________________________________

Since **px-vis-xy-chart** does not support **px-vis-markers** . Iam Creating this PR/feature to support  **px-vis-markers** in **px-vis-xy-chart** as similar to **px-vis-timeseries**. 

Below is the existing sample code of **px-vis-xy-chart** without markers followed by **px-vis-xy-chart** with markers along with the screenshots.

```
<px-vis-xy-chart
debounce-resize-timing="250"
width="1238.015625"
height="369.59375"
chart-horizontal-alignment="center"
chart-vertical-alignment="center"
margin='{"top":40,"bottom":120,"left":90,"right":20}'
tooltip-config='{"forceDateTimeDisplay":true}'
register-config='{"type":"horizontal","forceDateTimeDisplay":true}'
selection-type="xy"
chart-data="[[chartData]]"
series-config='{"firstSerie":{"type":"scatter","name":"Serie1","xAxisUnit":"Hz","yAxisUnit":"Hz","x":"T48A","y":"T48B","axis":{"id":"axis1","number":1,"side":"left"}},"secondSerie":{"type":"line","name":"Serie2","x":"T48A","y":"Line","axis":{"id":"axis2","number":2,"side":"left"}},"thirdSerie":{"type":"scatter","name":"Serie3","xAxisUnit":"Hz","yAxisUnit":"Hz","x":"T48A","y":"T48C","axis":{"id":"axis1","number":1,"side":"left"}}}'
chart-extents='{"x":["dynamic","dynamic"],"y":["dynamic","dynamic"]}'
event-data='[{"id":"123","label":"Fa Icon","x":"1334"},{"id":"456","label":"Unicode","x":"1342"},{"id":"789","label":"Default","x":"1346"}]'
event-config='{"Fa Icon":{"color":"blue","dataKey":"x","icon":"px-fea:deployments","type":"px","offset":[0,0]},"Unicode":{"color":"green","dataKey":"x","icon":"px-fea:bug","type":"px","offset":[1,0]},"Event Png":{"icon":"ge_logo.png","dataKey":"x","type":"image","offset":[0,-20],"size":"20"}}'

x-axis-config='{"title":"X","labelPosition":"center","orientation":"bottom"}'
y-axis-config='{"title":"Y","labelPosition":"center","orientation":"left"}'
time-data="Date"
dynamic-menu-config='[{"name":"Delete","action":"function(data) {var newConf = {}, confKeys = Object.keys(this.seriesConfig); for(var i=0; i<confKeys.length; i++) { if(data.additionalDetail.name !== confKeys[i]) {newConf[confKeys[i]] = this.seriesConfig[confKeys[i]];}}this.set(\"seriesConfig\", newConf);}","eventName":"delete","icon":"px-vis:trash-series"},{"name":"Bring To Front","action":"function(data) {this.set(\"serieToRedrawOnTop\", data.additionalDetail.name);}","eventName":"bring-to-front","icon":"px-vis:bring-to-front"}]'
toolbar-config='{"config":{"advancedZoom":true,"pan":true,"tooltip":true,"logHover":{"buttonGroup":2,"tooltipLabel":"The submenu item of this menu will define custom mouse interaction","icon":"px-nav:notification","subConfig":{"customClick":{"icon":"px-nav:play","buttonGroup":3,"tooltipLabel":"define some custom mouse interactions on chart","eventName":"my-custom-click","actionConfig":{"mousedown":"function(mousePos) { console.log(\"custom click on chart. Context is the chart. Mouse pos is available: \" + JSON.stringify(mousePos))}","mouseup":"function(mousePos) { console.log(\"custom action on mouse up the chart \" + JSON.stringify(mousePos));}","mouseout":"function(mousePos) { console.log(\"custom action on mouse out the chart \" + JSON.stringify(mousePos));}","mousemove":"function(mousePos) { console.log(\"custom action on hovering the chart \");}"}},"customClick2":{"buttonGroup":3,"icon":"px-nav:pause","tooltipLabel":"Remove all custom interactions","actionConfig":{"mousedown":null,"mouseup":null,"mouseout":null,"mousemove":null}}},"initialKey":"logHover"}}}'
nav-height="100"
nav-margin='{"top":5,"right":10,"bottom":20,"left":10}'
nav-series-config='{"T48A":{"type":"line","name":"Trend","x":"Date","y":"T48A"}}'>
</px-vis-xy-chart>
```
![image](https://user-images.githubusercontent.com/7873474/54790330-fb316d00-4bf2-11e9-85d7-d8408c6c3d7d.png)



px-vis-xy-chart with px-vis-markers.
**Added properties:** 
mark-data
marker-config

```
 <px-vis-xy-chart
debounce-resize-timing="250"
width="2078.0390625"
height="369.6328125"
chart-horizontal-alignment="center"
chart-vertical-alignment="center"
margin='{"top":40,"bottom":120,"left":90,"right":20}'
tooltip-config='{"forceDateTimeDisplay":true}'
register-config='{"type":"horizontal","forceDateTimeDisplay":true}'
selection-type="xy"
chart-data="[[chartData]]"
series-config='{"firstSerie":{"type":"scatter","name":"Serie1","xAxisUnit":"Hz","yAxisUnit":"Hz","x":"T48A","y":"T48B","axis":{"id":"axis1","number":1,"side":"left"}},"secondSerie":{"type":"line","name":"Serie2","x":"T48A","y":"Line","axis":{"id":"axis2","number":2,"side":"left"}},"thirdSerie":{"type":"scatter","name":"Serie3","xAxisUnit":"Hz","yAxisUnit":"Hz","x":"T48A","y":"T48C","axis":{"id":"axis1","number":1,"side":"left"}}}'
chart-extents='{"x":["dynamic","dynamic"],"y":["dynamic","dynamic"]}'
event-data='[{"id":"123","label":"Fa Icon","x":"1334"},{"id":"456","label":"Unicode","x":"1342"},{"id":"789","label":"Default","x":"1346"}]'
event-config='{"Fa Icon":{"color":"blue","dataKey":"x","icon":"px-fea:deployments","type":"px","offset":[0,0]},"Unicode":{"color":"green","dataKey":"x","icon":"px-fea:bug","type":"px","offset":[1,0]},"Event Png":{"icon":"ge_logo.png","dataKey":"x","type":"image","offset":[0,-20],"size":"20"}}'
marker-data='[{"id":"1","label":"Advisory","time":1505433900000,"datapoint":"1336"},{"id":"2","label":"Adaptation","time":1508025900000,"datapoint":"1338"},{"id":"3","label":"smooth","time":1510704300000,"datapoint":"1340"},{"id":"4","label":"Advisory2","time":1505433900000,"datapoint":"1344"},{"id":"5","label":"Adaptation2","time":1508025900000,"datapoint":"1348"},{"id":"6","label":"smooth2","time":1510704300000,"datapoint":"1350"}]'
marker-config='{"Advisory":{"location":"top","color":"rgb(180,140,242)","markerSymbol":"circle","priority":4,"markerSize":25,"markerScale":1,"markerFillOpacity":"1","markerStrokeOpacity":1,"row":0,"timezone":"UTC","firstDateTimeFormat":"hh:mm:ss A z"},"Adaptation":{"location":"bottom","color":"rgb(222,82,136)","markerSymbol":"circle","priority":5,"markerSize":34,"markerScale":1,"markerFillOpacity":"1","markerStrokeOpacity":1,"row":1,"timezone":"UTC","firstDateTimeFormat":"hh:mm:ss A z"},"smooth":{"location":"bottom","color":"rgb(222,82,136)","markerSymbol":"circle","priority":5,"markerSize":34,"markerScale":1,"markerFillOpacity":"1","markerStrokeOpacity":1,"row":1,"timezone":"UTC","firstDateTimeFormat":"hh:mm:ss A z"},"Advisory2":{"location":"bottom","color":"rgb(180,140,242)","markerSymbol":"square","priority":4,"markerSize":25,"markerScale":1,"markerFillOpacity":"1","markerStrokeOpacity":1,"row":1,"timezone":"UTC","firstDateTimeFormat":"hh:mm:ss A z"},"Adaptation2":{"location":"bottom","color":"rgb(222,82,136)","markerSymbol":"square","priority":5,"markerSize":34,"markerScale":1,"markerFillOpacity":"1","markerStrokeOpacity":1,"row":1,"timezone":"UTC","firstDateTimeFormat":"hh:mm:ss A z"},"smooth2":{"location":"bottom","color":"rgb(222,82,136)","markerSymbol":"square","priority":5,"markerSize":34,"markerScale":1,"markerFillOpacity":"1","markerStrokeOpacity":1,"row":1,"timezone":"UTC","firstDateTimeFormat":"hh:mm:ss A z"}}'
x-axis-config='{"title":"X","labelPosition":"center","orientation":"bottom"}'
y-axis-config='{"title":"Y","labelPosition":"center","orientation":"left"}'
time-data="Date"
dynamic-menu-config='[{"name":"Delete","action":"function(data) {var newConf = {}, confKeys = Object.keys(this.seriesConfig); for(var i=0; i<confKeys.length; i++) { if(data.additionalDetail.name !== confKeys[i]) {newConf[confKeys[i]] = this.seriesConfig[confKeys[i]];}}this.set(\"seriesConfig\", newConf);}","eventName":"delete","icon":"px-vis:trash-series"},{"name":"Bring To Front","action":"function(data) {this.set(\"serieToRedrawOnTop\", data.additionalDetail.name);}","eventName":"bring-to-front","icon":"px-vis:bring-to-front"}]'
toolbar-config='{"config":{"advancedZoom":true,"pan":true,"tooltip":true,"logHover":{"buttonGroup":2,"tooltipLabel":"The submenu item of this menu will define custom mouse interaction","icon":"px-nav:notification","subConfig":{"customClick":{"icon":"px-nav:play","buttonGroup":3,"tooltipLabel":"define some custom mouse interactions on chart","eventName":"my-custom-click","actionConfig":{"mousedown":"function(mousePos) { console.log(\"custom click on chart. Context is the chart. Mouse pos is available: \" + JSON.stringify(mousePos))}","mouseup":"function(mousePos) { console.log(\"custom action on mouse up the chart \" + JSON.stringify(mousePos));}","mouseout":"function(mousePos) { console.log(\"custom action on mouse out the chart \" + JSON.stringify(mousePos));}","mousemove":"function(mousePos) { console.log(\"custom action on hovering the chart \");}"}},"customClick2":{"buttonGroup":3,"icon":"px-nav:pause","tooltipLabel":"Remove all custom interactions","actionConfig":{"mousedown":null,"mouseup":null,"mouseout":null,"mousemove":null}}},"initialKey":"logHover"}}}'
nav-height="100"
nav-margin='{"top":5,"right":10,"bottom":20,"left":10}'
nav-series-config='{"T48A":{"type":"line","name":"Trend","x":"Date","y":"T48A"}}'>
</px-vis-xy-chart>

```
![image](https://user-images.githubusercontent.com/7873474/54790309-e6ed7000-4bf2-11e9-9dbb-31a3dff3e846.png)

saiDevarakonda commented:
HI Evan,

Updated the PR in expected format. Also updated the Demo.

As your tired before, download the branch and run in local. 
Note: include my px-vis and bower install.

Please let me know if you still have and questions or if you need more Details. 

Thank you.

Commit:
Fixed whitespace and comment formatting issues

Commit:
Updated bower.json

evanjd reviewed (approved):
Thanks for the explanation, makes sense.

I've made a few tweaks, will merge shortly.

evanjd commented:
@saiDevarakonda Change is live on predix-ui.com. px-vis-xy-chart [release v4.1.0](https://github.com/predixdesignsystem/px-vis-xy-chart/releases/tag/v4.1.0) includes your change. Cheers.
